sdk/python/kfp/client/auth.py

```diff
@@ -221,5 +341,91 @@ def id_token_from_refresh_token(client_id, client_secret, refresh_token,
         'audience': audience
     }
     res = requests.post(OAUTH_TOKEN_URI, data=payload)
-    res.raise_for_status()
-    return str(json.loads(res.text)[u'id_token'])
+    try:
+        res.raise_for_status()
+    except requests.exceptions.HTTPError as err:
+        raise ValueError(
+            ('Some HTTPErrors are caused by expired credentials in '
+             '~/.config/kfp/context.json. Try renaming or moving '
+             'them to another directory before trying again.')) from err
+    return str(json.loads(res.text)['id_token'])
+
+
+@contextmanager
+def get_local_server_app(host: str, port: int):
+    """Creates a local web-server for given host and port.
+
+    Args:
+        host (str): hostname of the server
+        port (int): port of the server
+    Returns:
+        WSGIServer: local server instance
+        RedirectWSGIApp: WSGI app to handle the authorization redirect
+    """
+    success_message = ('Kubeflow SDK authentication is completed.'
+                       ' You may close this window now.')
+    wsgi_app = RedirectWSGIApp(success_message)
+    wsgiref.simple_server.WSGIServer.allow_reuse_address = False
+    local_server = wsgiref.simple_server.make_server(
+        host,
+        port,
+        wsgi_app,
+        handler_class=wsgiref.simple_server.WSGIRequestHandler)
+    try:
+        yield local_server, wsgi_app
+    finally:
+        local_server.server_close()
+        del wsgi_app
+
+
+class RedirectWSGIApp:
+    """WSGI app to handle the authorization redirect.
+
+    Stores the request URI and displays the given success message.
+    """
+
+    def __init__(self, success_message):
+        """
+        Args:
+            success_message (str): The message to display in the web browser
+                the authorization flow is complete.
+        """
+        self.last_request_uri = None
+        self._success_message = success_message
+
+    def __call__(self, environ, start_response):
+        """WSGI Callable. Updates environment dictionary with parameters
+        required for WSGI and returns it.
+
+        Args:
+            environ (Mapping[str, Any]): The WSGI environment.
+            start_response (Callable[str, list]): The WSGI start_response
+                callable.
+        Returns:
+            Iterable[bytes]: The response body.
+        """
+        wsgiref.util.setup_testing_defaults(environ)
+        start_response('200 OK',
+                       [('Content-type', 'text/plain; charset=utf-8')])
+        self.last_request_uri = wsgiref.util.request_uri(environ)
+        return [self._success_message.encode('utf-8')]
+
+
+def fetch_auth_token_from_response(url: str):
+    """Fetches authorization code for OAuth2.0 Loopback flow.
+
+    Args:
+        url (str): a string containing the response URL
```

Comment:
nit: remove docstring types throughout. This information is redundant with type annotations.

Reply:
Thank you! Fixed